### DISCUSSION TITLE
[Feature request] Remote error reporting

### DISCUSSION THREAD

**Discussion by neuronix** [airsdk/Adobe-Runtime-Support#640] [Ideas]
In order to speed up bug fixing, I wonder if it would be possible to add an opt-in automatic error reporting service to AIR so that Harman servers can log errors (crashes & ANRs) with all the relevant device related.

Unfortunately the reality is that most developers don't properly track, examine & report errors on their apps so we have to rely on a handful of active devs reporting errors when they pick them up. Furthermore, many errors cannot be reproduced and/or cannot be understood with only a couple of error reports but may allow you to understand the issue and fix it if you have a large enough sample.

Best case would be for it to be opt-out (maybe only for licensed users as well) to have the largest reach possible.

Thanks

**Comment by henke37** [comment]
This is something that is best solved entirely within actionscript. And thanks to the UncaughtErrorEvents system, it is a breeze to implement.

All that is needed here is the service.

**Reply by neuronix** [reply]
No, I am referring to AIR SDK level errors, which are lower level. Save the crash reports and send them to Harman servers on next launch for example.

We already use UncaughtErrorEvents etc. for our application level reporting, the issue is Harman has no clue what errors are going around unless people painstakingly report them one by one and not only is this inefficient, it means that you only get error reports from a handful of users and their specific use cases.

It would allow Harman to fix issues that are widespread but under the radar and also monitor new SDK releases for new bugs and/or regressions.

**Reply by itlancer** [reply]
May be it could be some native extension that handle low-level errors/exceptions/crashes and provide event or interface to grab error data on next application launch.

**Reply by neuronix** [reply]
They already exist, distriqt provides one, but that is again application level stuff. I am aiming for tracking AIRSDK level errors to improve AIR's reliability.

**Reply by marchbold** [reply]
So you are basically talking about linking something like the [Exceptions ANE](https://github.com/distriqt/ANE-Exceptions) to an online service created by Harman? 

There is probably a lot of work involved in setting up a scalable web solution to handle crash reporting. I suppose it could be beneficial for Harman to be able to analyse crash reports from all AIR applications though. 

However I'm not sure why you wouldn't look into one of the many already available solutions, like Crashlytics etc? Generally they are super easy to integrate and produce really great data for developers to proactively respond to issues.

**Reply by neuronix** [reply]
No I am talking about adding a feature in the actual SDK. It has to be as low level as possible to track ANE related issues as well.

What seems unclear is that I am not asking for a feature for developers but for Harman themselves. So they can have as many detailed crash reports centralized so they can identify issues that are widespread without having to rely on developers reporting them one by one… we already all have crash reports on the Xcode and Google console without extra tools but unfortunately they do a very poor job of reporting all errors and can’t share the errors with Harman.

We built such a tool 8 years ago in about a day (dump, DB, statistics) for our apps on an application level, with all the scaling cloud services available, the main obstacle isn’t technical but legal.. with GDPR and such, opt-out wouldn’t be possible so adoption would not that huge but if at least a few of the biggest developers opted-in it would provide very valuable data for harman to provide more reliable SDKs.

**Reply by marchbold** [reply]
Yeah for sure, definitely would be valuable for Harman, but the implementation would be as I described whether it was internal to the SDK or not.

However I'd still encourage it to be separate from the core SDK. Reason being is that generally you can only have one crash reporting service implemented as they conflict with each other (basically cannot have multiple unhandled exception handlers). So if we wanted to use a different service there wouldn't be any use of this code in the AIR SDK.

The majority of the work is still in developing a web service to receive the reports as the implementation in the exceptions ane is all you would need on a device apart from a small addition to send it to the web service.

**Reply by neuronix** [reply]
I didn’t know that but it makes sense. I don’t know how feasible the idea really is, it just has been such a valuable tool for us in house on the application level, I think it would be great for the AIRSDK if possible.

**Reply by marchbold** [reply]
For sure, it's a great idea!
